src/h3/frame.rs

```diff
@@ -0,0 +1,879 @@
+        let frame = match frame_type {
+            DATA_FRAME_TYPE_ID => Frame::Data {
+                payload: b.get_bytes(payload_length as usize)?.to_vec(),
+            },
+
+            HEADERS_FRAME_TYPE_ID => Frame::Headers {
+                header_block: b.get_bytes(payload_length as usize)?.to_vec(),
+            },
+
+            PRIORITY_FRAME_TYPE_ID => parse_priority_frame(b)?,
+
+            CANCEL_PUSH_FRAME_TYPE_ID => Frame::CancelPush {
+                push_id: b.get_varint()?,
+            },
+
+            SETTINGS_FRAME_TYPE_ID => parse_settings_frame(payload_length, b)?,
+
+            PUSH_PROMISE_FRAME_TYPE_ID => parse_push_promise(payload_length, b)?,
+
+            GOAWAY_FRAME_TYPE_ID => Frame::GoAway {
+                stream_id: b.get_varint()?,
+            },
+
+            MAX_PUSH_FRAME_TYPE_ID => Frame::MaxPushId {
+                push_id: b.get_varint()?,
+            },
+
+            DUPLICATE_PUSH_FRAME_TYPE_ID => Frame::DuplicatePush {
+                push_id: b.get_varint()?,
+            },
+
+            _ => return Err(Error::InvalidFrame),
+        };
+
+        Ok(frame)
+    }
+
+    pub fn from_bytes2(
+        frame_type: u8, payload_length: u64, bytes: &mut [u8],
+    ) -> Result<Frame> {
+        // let payload_length = b.get_varint()?;
+        // let frame_type = b.get_u8()?;
+
+        // debug!("GOT FRAME {:x}, payload_len= {}", frame_type, payload_length);
+        // dbg!(&bytes);
+
+        let mut b = octets::Octets::with_slice(bytes);
+
+        // TODO: handling of 0-length frames
+        let frame = match frame_type {
+            0x0 => Frame::Data {
+                payload: b.get_bytes(payload_length as usize)?.to_vec(),
+            },
+
+            0x1 => Frame::Headers {
+                header_block: b.get_bytes(payload_length as usize)?.to_vec(),
+            },
+
+            0x02 => parse_priority_frame(&mut b)?,
+
+            0x03 => Frame::CancelPush {
+                push_id: b.get_varint()?,
+            },
+
+            0x04 => parse_settings_frame(payload_length, &mut b)?,
+
+            0x05 => parse_push_promise(payload_length, &mut b)?,
+
+            0x07 => Frame::GoAway {
+                stream_id: b.get_varint()?,
+            },
+
+            0x0D => Frame::MaxPushId {
+                push_id: b.get_varint()?,
+            },
+
+            0x0E => Frame::DuplicatePush {
+                push_id: b.get_varint()?,
+            },
+            _ => {
+                error!("Invalid frame type {}", frame_type);
+                return Err(Error::InvalidFrame);
+            },
+        };
+
+        Ok(frame)
+    }
+
+    pub fn to_bytes(&self, b: &mut octets::Octets) -> Result<usize> {
+        let before = b.cap();
+
+        trace!("Serializing frame type {:?}", self);
+
+        match self {
+            Frame::Data { payload } => {
+                b.put_varint(payload.len() as u64)?;
+                b.put_varint(0x00)?;
```

Comment:
Should probably use the frame ID constants from above since you went to the trouble of defining them :)